File: P2.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 27 13:02:47 2021

@author: jgdch
"""

#Time Compleixty:O(N), N is the number of ndoes in the tree
#Space Complexity:O(H), H is the height of the tree
#Passed all LC testcases succesfully

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.isSym(root,root)
    def isSym(self,root,root1):
        if(root==None and root1==None):
            return True
        if(root==None or root1==None):
            return False
        return (root.val==root1.val) and self.isSym(root.right,root1.left) and self.isSym(root.left,root1.right)
    
        

File: p2.java
// Time Complexity : O(nodes)
// Space Complexity : O(nodes)
// Did this code successfully run on Leetcode : Y
// Any problem you faced while coding this : N


// Your code here along with comments explaining your approach
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean flag = false;
    public boolean isSymmetric(TreeNode root) {
        if (root==null) return true;
        flag = true;
        check_sym(root.left, root.right);
        return flag;
    }
    private void check_sym(TreeNode left, TreeNode right)
    {
        
        if(left==null && right==null)
        {
            return;
        }
        if(left==null || right==null || left.val!=right.val)
        {
            flag = false;
            return;
        }
        check_sym(left.left, right.right);
        check_sym(left.right, right.left);
    }
}








